Forwarded: https://github.com/MariaDB/server/pull/2555
Origin: https://patch-diff.githubusercontent.com/raw/MariaDB/server/pull/2555.patch
From: Christian Gonzalez <gondchri@amazon.com>
Date: Wed, 15 Mar 2023 23:29:46 +0000
Subject: [PATCH] MDEV-30483 Add fix for migration from MySQL 5.7 to MariaDB
 10.6

As described in MDEV-30483 and MDEV-30809 the migration path from MySQL
5.7 to MariaDB 10.6 is broken because the PRTYPE of `last_update` column
for tables `mysql.innodb_table_stats` and `mysql.innodb_index_stats` is
not being properly updated after running the corresponding alter table
command. This commit adds logic into the `scripts/mysql_system_tables_fix.sql`
script to force the tables rebuild when the alter table does not update
the `PRTYPE` properly by using the `force` flag, after the table rebuild
the `PRTYPE` is properly updated.

All new code of the whole pull request, including one or several files
that are either new files or modified ones, are contributed under the
BSD-new license. I am contributing on behalf of my employer
Amazon Web Services, Inc.
---
 scripts/mysql_system_tables_fix.sql | 19 +++++++++++++++++++
 1 file changed, 19 insertions(+)

--- a/scripts/mysql_system_tables_fix.sql
+++ b/scripts/mysql_system_tables_fix.sql
@@ -756,8 +756,27 @@ if @have_innodb then
   delete from innodb_table_stats where length(table_name) > 64;
 
   # update table_name and timestamp fields in the innodb stat tables
+  #
+  # FIXME: after running the first alter table statement, the value of PRTYPE is
+  # checked against the PRTYPE value of a fresh installation. For some reason
+  # after migrating from MySQL 5.7 the PRTYPE is not being properly updated and
+  # the force flag is required to rebuild the table and force the update of the
+  # PRTYPE when this happens. More information available at MDEV-30483 and MDEV-30809
   alter table innodb_index_stats modify last_update timestamp not null default current_timestamp on update current_timestamp, modify table_name varchar(199);
+  set @prtype= (select b.PRTYPE != 526087
+                from information_schema.innodb_sys_tables a, information_schema.innodb_sys_columns b
+                where a.table_id=b.table_id and a.name = 'mysql/innodb_index_stats' and b.name = 'last_update');
+  if @prtype then
+    alter table innodb_index_stats modify last_update timestamp not null default current_timestamp on update current_timestamp, force;
+  end if;
+
   alter table innodb_table_stats modify last_update timestamp not null default current_timestamp on update current_timestamp, modify table_name varchar(199);
+   set @prtype= (select b.PRTYPE != 526087
+                 from information_schema.innodb_sys_tables a, information_schema.innodb_sys_columns b
+                 where a.table_id=b.table_id and a.name = 'mysql/innodb_table_stats' and b.name = 'last_update');
+  if @prtype then
+    alter table innodb_table_stats modify last_update timestamp not null default current_timestamp on update current_timestamp, force;
+  end if;
 
   alter table innodb_index_stats drop foreign key if exists innodb_index_stats_ibfk_1;
 end if //
